streams/src/test/java/org/apache/kafka/streams/tests/EosTestClient.java

```diff
@@ -0,0 +1,150 @@
+import org.apache.kafka.streams.KafkaStreams;
+import org.apache.kafka.streams.StreamsConfig;
+import org.apache.kafka.streams.kstream.Aggregator;
+import org.apache.kafka.streams.kstream.Initializer;
+import org.apache.kafka.streams.kstream.KGroupedStream;
+import org.apache.kafka.streams.kstream.KStream;
+import org.apache.kafka.streams.kstream.KStreamBuilder;
+
+import java.io.File;
+import java.util.Properties;
+import java.util.concurrent.TimeUnit;
+
+public class EosTestClient extends SmokeTestUtil {
+
+    static final String APP_ID = "EosTest";
+    private final String kafka;
+    private final File stateDir;
+    private KafkaStreams streams;
+    private boolean uncaughtException;
+
+    EosTestClient(final File stateDir, final String kafka) {
+        super();
+        this.stateDir = stateDir;
+        this.kafka = kafka;
+    }
+
+    private boolean isRunning = true;
+
+    public void start() {
+        Runtime.getRuntime().addShutdownHook(new Thread(new Runnable() {
+            @Override
+            public void run() {
+                isRunning = false;
+                streams.close(5, TimeUnit.SECONDS);
+                // do not remove these printouts since they are needed for health scripts
+                if (!uncaughtException) {
+                    System.out.println("EOS-TEST-CLIENT-CLOSED");
+                }
+            }
+        }));
+
+        while (isRunning) {
+            if (streams == null) {
+                uncaughtException = false;
+
+                streams = createKafkaStreams(stateDir, kafka);
+                streams.setUncaughtExceptionHandler(new Thread.UncaughtExceptionHandler() {
+                    @Override
+                    public void uncaughtException(final Thread t, final Throwable e) {
+                        System.out.println("EOS-TEST-CLIENT-EXCEPTION");
+                        e.printStackTrace();
+                        uncaughtException = true;
+                    }
+                });
+                streams.start();
+            }
+            if (uncaughtException) {
+                streams.close(5, TimeUnit.SECONDS);
+                streams = null;
+            }
+            sleep(1000);
+        }
+    }
+
+    private static KafkaStreams createKafkaStreams(final File stateDir,
+                                                   final String kafka) {
+        final Properties props = new Properties();
+        props.put(StreamsConfig.APPLICATION_ID_CONFIG, APP_ID);
+        props.put(StreamsConfig.STATE_DIR_CONFIG, stateDir.toString());
+        props.put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, kafka);
+        props.put(StreamsConfig.NUM_STREAM_THREADS_CONFIG, 2);
+        props.put(StreamsConfig.NUM_STANDBY_REPLICAS_CONFIG, 2);
+        props.put(StreamsConfig.REPLICATION_FACTOR_CONFIG, 3);
+        props.put(StreamsConfig.PROCESSING_GUARANTEE_CONFIG, StreamsConfig.EXACTLY_ONCE);
+        props.put(StreamsConfig.CACHE_MAX_BYTES_BUFFERING_CONFIG, 0);
```

Comment:
With caching, we cannot know the result (ie, not each individual record that get's written to the output), as we don't know when we will flush. For this test, I want to check all intermediate records, too, so I need to disable caching.

But we can add a second test with caching and just make sure we don't over count for example and ignore all intermediate results and just check the latest result per key.